lib/icinga/clusterevents.cpp

```diff
@@ -949,5 +949,47 @@ Value ClusterEvents::ExecutedCommandAPIHandler(const MessageOrigin::Ptr& origin,
 
 Value ClusterEvents::UpdateExecutionsAPIHandler(const MessageOrigin::Ptr& origin, const Dictionary::Ptr& params)
 {
-	return "Not implemented";
-}
+	Endpoint::Ptr endpoint = origin->FromClient->GetEndpoint();
+	if (!endpoint) {
+		Log(LogNotice, "ClusterEvents")
+				<< "Discarding 'update executions API handler' message from '" << origin->FromClient->GetIdentity()
+				<< "': Invalid endpoint origin (client not allowed).";
+
+		return Empty;
+	}
+
+	Host::Ptr host = Host::GetByName(params->Get("host"));
+	if (!host)
+		return Empty;
+
+	Checkable::Ptr checkable;
+
+	if (params->Contains("service"))
+		checkable = host->GetServiceByShortName(params->Get("service"));
+	else
+		checkable = host;
+
+	if (!checkable)
+		return Empty;
+
+	ObjectLock oLock (checkable);
+
+	if (origin->FromZone && !origin->FromZone->CanAccessObject(checkable)) {
+		Log(LogNotice, "ClusterEvents")
+				<< "Discarding 'update executions API handler' message for checkable '" << checkable->GetName()
+				<< "' from '" << origin->FromClient->GetIdentity() << "': Unauthorized access.";
+		return Empty;
+	}
+
+	if (!params->Contains("executions")) {
+		Log(LogNotice, "ClusterEvents")
+				<< "Discarding 'update executions API handler' message for checkable '" << checkable->GetName()
+				<< "' from '" << origin->FromClient->GetIdentity() << "': missing executions.";
+		return Empty;
+	}
+
+	Dictionary::Ptr executions = params->Get("executions");
+	checkable->SetExecutions(executions);
+
+	return Empty;
+}
```

Comment:
Find this comment of mine in the files changed tab, not in the conversation one. Do you see the round red sign?